crates/bevy_asset/src/path.rs

```diff
@@ -63,6 +63,51 @@ impl<'a> AssetPath<'a> {
                 .map(|value| Cow::Owned(value.to_string())),
         }
     }
+
+    /// Resolves a relative asset path. The result will be an `AssetPath` which is resolved relative
+    /// to this path. There are three cases:
+    ///
+    /// If the relative path begins with `#`, then it is considered an asset label, in which case
+    /// the result is the base path with the label portion replaced.
+    ///
+    /// If the relative path starts with "./" or "../", then the result is combined with the base
+    /// path. The rules are the same as for URIs, which means that the relative path is resolved
+    /// relative to the *directory* of the base path, so a base path of `"x/y/z#foo"` combined with
+    /// a relative path of `"./a#bar"` yields `"x/y/a#bar"`.
+    ///
+    /// If neither of the above are true, then the relative path is considered a 'full' path,
+    /// and the result is simply a copy of the relative path, converted into an `AssetPath`.
+    /// Note that a 'full' asset path is still relative to the asset root directory, and not
+    /// necessarily an absolute filesystem path.
+    pub fn resolve(&'a self, relative_path: &'a str) -> AssetPath<'a> {
+        if let Some(without_prefix) = relative_path.strip_prefix('#') {
+            AssetPath::new_ref(&self.path, Some(without_prefix))
+        } else if relative_path.starts_with("./") || relative_path.starts_with("../") {
+            let mut rpath = relative_path;
+            let mut fpath = PathBuf::from(self.path());
+            if !fpath.pop() {
+                panic!("Can't compute relative path - not enough path elements");
+            }
+            loop {
+                if let Some(rest) = rpath.strip_prefix("./") {
+                    rpath = rest;
+                } else if let Some(rest) = rpath.strip_prefix("../") {
+                    rpath = rest;
+                    if !fpath.pop() {
+                        panic!("Can't compute relative path - not enough path elements");
+                    }
+                } else {
+                    break;
+                }
```

Comment:
I don't think this works for paths with multiple `./` or `../`. For example, `AssetPath::resolve("../foo/../bar")` doesn't result in the correct path.

I used [`path_clean`](https://crates.io/crates/path-clean) for this, but you could alternatively roll the logic out yourself.

It might seem odd that this would ever come up, but I've seen it before with generated paths.

Reply:
Yeah, I was wondering whether or not to canonicalize internal parent indicators. That makes the logic more complex, and (given the Bevy mandate for minimizing dependencies) I don't want to bring in an additional crate for this. I'll take a look and see what I can do.

Reply:
Done, added additional canonicalization and tests.

Note that the base path is not canonicalized, only the parts that were appended.